lib/AST/ClangTypeConverter.cpp

```diff
@@ -0,0 +1,684 @@
+        if (!NSUIntegerTy.isNull()) {
+          Cache.insert({swiftType->getDesugaredType(), NSUIntegerTy});
+          return;
+        }
+      }
+
+      // For something like `typealias CInt = Int32`, reverseBuiltinTypeMapping
+      // will get Int32 as the input, so we need to record the desugared type.
+      Cache.insert({swiftType->getDesugaredType(),
+                    getClangBuiltinTypeFromKind(ctx, builtinKind)});
+    };
+
+#define MAP_BUILTIN_TYPE(CLANG_BUILTIN_KIND, SWIFT_TYPE_NAME)          \
+    cacheStdlibType(#SWIFT_TYPE_NAME, clang::BuiltinType::CLANG_BUILTIN_KIND);
+#include "swift/ClangImporter/BuiltinMappedTypes.def"
+
+    // On 64-bit Windows, no C type is imported as an Int or UInt; CLong is
+    // imported as an Int32 and CLongLong as an Int64. Therefore, manually
+    // add mappings to C for Int and UInt.
+    if (Triple.isOSWindows() && Triple.isArch64Bit()) {
+      // Map UInt to uintptr_t
+      auto swiftUIntType = getNamedSwiftType(stdlib, "UInt");
+      auto clangUIntPtrType = ctx.getCanonicalType(ctx.getUIntPtrType());
+      Cache.insert({swiftUIntType, clangUIntPtrType});
+
+      // Map Int to intptr_t
+      auto swiftIntType = getNamedSwiftType(stdlib, "Int");
+      auto clangIntPtrType = ctx.getCanonicalType(ctx.getIntPtrType());
+      Cache.insert({swiftIntType, clangIntPtrType});
+    }
+    StdlibTypesAreCached = true;
+  }
+
+  auto it = Cache.find(Type(type));
+  if (it != Cache.end())
+    return it->second;
+
+  it = Cache.find(type->getCanonicalType());
+  if (it != Cache.end()) {
+    Cache.insert({Type(type), it->second});
+    return it->second;
+  }
+
+  return clang::QualType();
+}
+
+clang::QualType ClangTypeConverter::visitTupleType(TupleType *type) {
+  unsigned tupleNumElements = type->getNumElements();
+  if (tupleNumElements == 0)
+    return ClangASTContext.VoidTy;
+
+  Type eltTy = type->getElementType(0);
+  for (unsigned i = 1; i < tupleNumElements; i++) {
+    if (!eltTy->isEqual(type->getElementType(i)))
+      // Only tuples where all element types are equal map to fixed-size
+      // arrays.
+      return clang::QualType();
+  }
+
+  auto clangEltTy = convert(eltTy);
+  if (clangEltTy.isNull())
+    return clang::QualType();
+
+  APInt size(32, tupleNumElements);
+  return ClangASTContext.getConstantArrayType(clangEltTy, size,
+           clang::ArrayType::Normal, 0);
+}
+
+clang::QualType ClangTypeConverter::visitProtocolType(ProtocolType *type) {
+  auto proto = type->getDecl();
+
+  // Single protocol -> id<Proto>
+  if (proto->isObjC()) {
+    auto &clangCtx = ClangASTContext;
+    clang::IdentifierInfo *name = &clangCtx.Idents.get(proto->getName().get());
+    auto *PDecl = clang::ObjCProtocolDecl::Create(
+                    const_cast<clang::ASTContext &>(clangCtx),
+                    clangCtx.getTranslationUnitDecl(), name,
+                    clang::SourceLocation(), clang::SourceLocation(), nullptr);
```

Comment:
1. How would I check the import?
2. “Re-use the fake protocol” — won’t that happen automatically because ClangTypeConverter caches things anyways? Or is this about the fact that we don’t cache sub-computations?

Reply:
We discussed this privately, but just for a public answer: for the first, you can check whether the Swift Decl has a clang node, which should be a `clang::ObjCProtocolDecl*`; for the second, you should endeavor to make sure that different Swift types use the same Clang declaration, both to keep down the memory use and to make this more useful as we find more clients.